include/seqan3/argument_parser/validators.hpp

```diff
@@ -0,0 +1,386 @@
+ * exception whenever a given value is not in the given list.
+ *
+ *```cpp
+ *     int main(int argc, char ** argv)
+ *     {
+ *          seqan3::argument_parser myparser(argc, argv); // initialize
+ *
+ *          int myint;
+ *          seqan3::value_list_validator my_validator{2,4,6,8,10};
+ *
+ *          myparser.add_option(myint,'i',"integer","Give me a number.",
+ *                              seqan3::option_spec::DEFAULT, my_validator);
+ *
+ *          // an exception will be thrown if the user specifies an integer
+ *          // that is not one of [2,4,6,8,10] (e.g. "./test_app -i 3")
+ *          try
+ *          {
+ *               myparser.parse();
+ *          }
+ *          catch (seqan3::parser_invalid_argument const & ext) // the user did something wrong
+ *          {
+ *               std::cerr << "[PARSER ERROR] " << ext.what(); // customize your error message
+ *               return -1;
+ *          }
+ *          catch (seqan3::parser_interruption const &) // expected behaviour on special requests (e.g. `--help`)
+ *          {
+ *               return 0;
+ *          }
+ *
+ *          std::cout << "integer given by user passed validation: " << myint << endl;
+ *          return 0;
+ *     }
+ *```
+ */
+template <typename option_value_type>
+class value_list_validator
+{
+public:
+    using value_type = option_value_type; //! Type of values that are tested by validator
+
+    /*!\brief Constructing from a vector.
+     *
+     * \param v The vector of valid values to test.
+     */
+    value_list_validator(std::vector<option_value_type> const & v) :
+        values{v}
+    {}
+
+    /*!\brief Constructing from an initializer_list.
+     *
+     * \param v The initializer_list of valid values to test.
+     */
+    value_list_validator(std::initializer_list<option_value_type> const & v) :
+        values{v}
+    {}
+
+    /*!\brief Tests whether cmp lies inside values.
+     * \throws parser_invalid_argument
+     */
+    void operator()(option_value_type const & cmp) const
+    {
+        if (!(std::find(values.begin(), values.end(), cmp) != values.end()))
+            throw parser_invalid_argument(detail::paste("Value ", cmp, " is not one of ", ranges::view::all(values), "."));
+    }
+
+    //!\brief Tests whether every value of v lies inside values.
+    template <typename container_type>
+    //!\cond
+        requires container_concept<container_type> &&
+                 std::is_same<typename container_type::value_type, option_value_type>::value
+    //!\endcond
+    void operator()(container_type const & v) const
+    {
+        std::for_each(v.begin(), v.end(), [&] (auto cmp) { (*this)(cmp); });
+    }
+
+    //!\brief Returns a message that can be appended to the (positional) options help page info.
+    std::string get_help_page_message() const
+    {
+        return detail::paste("Value must be one of ", ranges::view::all(values), ".");
```

Comment:
what happens here?

Reply:
detail::paste() will stream all parameters into a string stream and return an std::string. I use the << operator of the range library to conveniently print the vector as a comma separated list. :smile: 
Is that ok?

Reply:
yes, but maybe rename detail::paste to `concat_as_string`?